File: 0112/dfs.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def hasPathSum(self, root: TreeNode, sum: int) -> bool:
        return self.recursion(root, 0, sum)

    def recursion(self, node, prev_sum, sum):
        if not node:
            return False
        curr_sum = prev_sum + node.val
        if not node.left and not node.right and curr_sum == sum:
            return True
        return self.recursion(node.left, curr_sum, sum) or \
            self.recursion(node.right, curr_sum, sum)


if __name__ == '__main__':
    sol = Solution()

    root = TreeNode(5)
    root.left = TreeNode(4)
    root.left.left = TreeNode(11)
    root.left.left.left = TreeNode(7)
    root.left.left.right = TreeNode(2)
    root.right = TreeNode(8)
    root.right.left = TreeNode(13)
    root.right.right = TreeNode(4)
    root.right.right.right = TreeNode(1)
    assert sol.hasPathSum(root, 22)

File: 0102/dfs.cpp
#include <cassert>
#include <iostream>
#include <limits>
#include <vector>

#include "leetcode/tree.hpp"

using namespace std;
using namespace LeetCode;

class Solution {
 public:
  vector<vector<int>> levelOrder(TreeNode* root) {
    vector<vector<int>> ret;
    dfs(root, ret, 0);
    return ret;
  }

  void dfs(TreeNode* node, vector<vector<int>>& ret, int level) {
    if (!node) return;
    if (level >= ret.size()) ret.resize(level+1);
    ret[level].push_back(node->val);
    dfs(node->left, ret, level+1);
    dfs(node->right, ret, level+1);
  }
};

int main() {
  Solution sol;
  TreeNode* node;
  vector<vector<int>> got;
  vector<int> expected;

  node = CreateTreeNode("3 9 20 # # 15 7");
  got = sol.levelOrder(node);
  expected = {3};
  assert (got[0] == expected);
  expected = {9,20};
  assert (got[1] == expected);
  expected = {15,7};
  assert (got[2] == expected);
  DeleteTreeNode(node);

  node = CreateTreeNode("1 2 3 4 5 6 7");
  got = sol.levelOrder(node);
  expected = {1};
  assert (got[0] == expected);
  expected = {2,3};
  assert (got[1] == expected);
  expected = {4,5,6,7};
  assert (got[2] == expected);
  DeleteTreeNode(node);

  got = sol.levelOrder(nullptr);
  assert(got.size() == 0);
}
